lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+         is an unambiguously-determined settable, and whether
+         an explicit request was necessary to disambiguate *)
+      let rec split rev_before : 'a t list -> _ = function
+        | [] -> assert false (* we know there is at least one choice *)
+        | c :: rest ->
+          if c.has_tmc_calls && (not explicit || is_explicit c) then
+            { rev_before; choice = c; after = List.map direct rest }
+          else
+            split (direct c :: rev_before) rest
+      in split [] choices
+    in
+    let tmc_call_subterms =
+      List.filter (fun c -> c.has_tmc_calls) choices
+    in
+    match tmc_call_subterms with
+    | [] ->
+        No_tmc_call (List.map direct choices)
+    | [ _one ] ->
+        Nonambiguous (nonambiguous ~explicit:false choices)
+    | several_subterms ->
+        let explicit_subterms = List.filter is_explicit several_subterms in
+        begin match explicit_subterms with
+        | [] -> Ambiguous several_subterms
+        | [ _one ] ->
+            Nonambiguous (nonambiguous ~explicit:true choices)
+        | several_explicit_subterms -> Ambiguous several_explicit_subterms
+        end
+end
+
+open Choice.Syntax
+
+type context = {
+  specialized: specialized Ident.Map.t;
+}
+and specialized = {
+  arity: int;
+  dps_id: Ident.t;
+}
+
+let find_candidate = function
+  | Lfunction lfun when lfun.attr.tmc_candidate -> Some lfun
+  | _ -> None
+
+let declare_binding ctx (var, def) =
+  match find_candidate def with
+  | None -> ctx
+  | Some lfun ->
+  let arity = List.length lfun.params in
+  let dps_id = Ident.create_local (Ident.name var ^ "_dps") in
+  let cand = { arity; dps_id } in
+  { specialized = Ident.Map.add var cand ctx.specialized }
+
+let rec choice ctx t =
+  let rec choice ctx ~tail t =
+    match t with
+    | (Lvar _ | Lconst _ | Lfunction _ | Lsend _
+      | Lassign _ | Lfor _ | Lwhile _) ->
+        let t = traverse ctx t in
+        Choice.return t
+
+    (* [choice_prim] handles most primitives, but the important case of construction
+       [Lprim(Pmakeblock(...), ...)] is handled by [choice_makeblock] *)
+    | Lprim (prim, primargs, loc) ->
+        choice_prim ctx ~tail prim primargs loc
+
+    (* [choice_apply] handles applications, in particular tail-calls which
+       generate Set choices at the leaves *)
+    | Lapply apply ->
+        choice_apply ctx ~tail apply
+    (* other cases use the [lift] helper that takes the sub-terms in tail
+       position and the context around them, and generates a choice for
+       the whole term from choices for the tail subterms. *)
+    | Lsequence (l1, l2) ->
+        let l1 = traverse ctx l1 in
+        let+ l2 = choice ctx ~tail l2 in
+        Lsequence (l1, l2)
+    | Lifthenelse (l1, l2, l3) ->
+        let l1 = traverse ctx l1 in
+        let+ (l2, l3) = choice_pair ctx ~tail (l2, l3) in
+        Lifthenelse (l1, l2, l3)
+    | Llet (lk, vk, var, def, body) ->
+        (* non-recursive bindings are not specialized *)
```

Comment:
Non recursive bindings should also be specialized for cases like this:
```OCaml
  type 'a t =
    | N of 'a
    | C of 'a t * 'a t

  let[@tail_mod_cons] rec map f l =
    match l with
    | N v -> N (f v)
    | C (a, b) ->
        let[@tail_mod_cons] map' l = map f l in
        C (map' a, (map' [@tailcall]) b)
```
I don't think that there are any downsides to it